### PR TITLE
GEODE-3967: when ConcurrentCacheModificationException happened. Gatew…

### PR DESCRIPTION
**Pull request by gesterzhou**
…aySenderEventImpl should save the status and notify gatewaysender anyway. SerialGatewaySender will handle it. In AbstractUpdateOperation's doPutOrCreate's 3 tries of basicUpdate, the 3rd one should allow both create and update.

One more bug I fixed is: for UPDATE_VERSION_STAMP event, SerialGatewaySender only send when it's primary.

@boglesby @dschneider-pivotal 

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java [hunk 1]
```diff
@@ -2840,6 +2840,8 @@ protected Object findObjectInSystem(KeyInfo keyInfo, boolean isCreate, TXStateIn
               logger.debug("caught concurrent modification attempt when applying {}", event);
             }
             notifyBridgeClients(event);
+            notifyGatewaySender(event.getOperation().isUpdate() ? EnumListenerEvent.AFTER_UPDATE
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java]: this one seems like it is not needed

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java [hunk 2]
```diff
@@ -171,6 +171,8 @@
 
   protected boolean isInitialized;
 
+  public boolean isConcurrencyConflict = false;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: Don't make fields public. Make it private and add public gettor/settor methods

File: geode-core/src/test/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java [hunk 3]
```diff
@@ -476,7 +476,9 @@ public void run() {
       CCRegion.basicBridgePut("cckey0", "newvalue", null, true, null, id, true, holder);
       vm0.invoke(new SerializableRunnable("check conflation count") {
         public void run() {
-          assertEquals("expected one conflated event", 1,
+          // after changed the 3rd try of AUO.doPutOrCreate to be ifOld=false ifNew=false
+          // ARM.updateEntry will be called one more time, so there will be 2 conflacted events
+          assertEquals("expected one conflated event", 2,
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: change message to be "two" conflated events

File: geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java [hunk 4]
```diff
@@ -1185,6 +1185,7 @@ public boolean destroy(EntryEventImpl event, boolean inTokenMode, boolean during
                               true/* conflict with clear */, duringRI, true);
                           doPart3 = true;
                         } catch (ConcurrentCacheModificationException ccme) {
+                          event.isConcurrencyConflict(true);
```
[the remark this answers is not in the document]
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java (hunk 1 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java (hunk 2 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: done

**Pull request closed without merging**